### PR TITLE
[SYCL] Add has_known_identity/known_identity

### PR DESCRIPTION
**Pull request by Pennycook**
These traits have been requested by developers who would like to
determine at compile-time whether the reduction() interface supports
their use-case, or whether they need to fall back to some other
implementation.

Exposes existing functionality of reduction implementation as part of
the public API for reductions.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (keryell) [inline sycl/doc/extensions/Reduction/Reduction.md:57]: What is `T`?

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (rolandschulz) [inline sycl/doc/extensions/Reduction/Reduction.md:44]: Why do we need this? Isn't this the same as `is_detected<known_identity, BinaryOperation, AccumulatorT>`?

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (rolandschulz) [reply sycl/doc/extensions/Reduction/Reduction.md:57]: Typo. Should be `AccumulatorT`.

Comment (Pennycook) [reply sycl/doc/extensions/Reduction/Reduction.md]: Oops.  Thanks, both: fixed in c2ff9dcde43c.

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (Pennycook) [reply sycl/doc/extensions/Reduction/Reduction.md:44]: @rarutyun argued that because `is_detected` isn't standard, it's nicer to expose a dedicated trait for this.

The reason this is different to the group algorithms fix is that non-expert users may still need to care about the identity values.  Reasoning about which form of the `reduction()` function to use and whether there may be some performance overhead is useful to application developers, not just library developers building an extra level of abstraction.

**Comment by Pennycook** [comment]
@Fznamznon, @AlexeySachkov: Could you please take a look at the build failures here and help identify a path forwards?

It looks to me like `reduction_nd_ext_half.cpp` is failing now because the calculation of negative infinity isn't `constexpr` for `half`, but I'm not sure how to fix it.  Is there an alternative way to write the below that is `constexpr` for `half`, or do we need to add some new `constexpr` functions to the host implementation of `half`?
```c++
  static constexpr AccumulatorT value =
      std::numeric_limits<AccumulatorT>::has_infinity
          ? static_cast<AccumulatorT>(
                -std::numeric_limits<AccumulatorT>::infinity())
          : std::numeric_limits<AccumulatorT>::lowest();
```

I think unary `-` triggers a conversion to `float` which isn't `constexpr`.  I tried making `operator float()` `constexpr`, but there's no host definition of this function that I could see.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (v-klochkov) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:182]: Minor comment, definitely not a request to change.
This could be a little bit shorter (but same number of lines though):
```suggestion
                           enable_if_t<IsMinimumIdentityOp<
                               AccumulatorT, BinaryOperation>::value>> {
```

**Review by v-klochkov — COMMENTED** [review]
Looks good to me, I found only 1 typo.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (v-klochkov) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:197]: I believe, adding static_cast would help to fix the failed test.
Please see the code before your changes. That additional static_cast was added mostly for Float16:
  getIdentity() {	
    return std::numeric_limits<_T>::has_infinity	
               ? static_cast<_T>(-std::numeric_limits<_T>::infinity())	
               : std::numeric_limits<_T>::lowest();	
  }

Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp:197]: Sorry, I did not notice, additional static_cast was already in place. Deleting my previous comment now.

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (Pennycook) [reply sycl/doc/extensions/Reduction/Reduction.md]: Oops.  Fixed in aa2222abdc047789114542b011af27158474d9bd.  Thanks!

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp:182]: I tried this during development and couldn't get it to work.  I chalked it up to me not understanding enough about how our implementation of `enable_if_t` differs from `std::enable_if`.  Can you compile locally if you make this change?  Maybe I'm doing something wrong.

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (rolandschulz) [reply sycl/doc/extensions/Reduction/Reduction.md:44]: I disagree. Not every problem needs to be solvable with just just SYCL and std. Users should use extra libraries for utilities.

Comment (rolandschulz) [reply sycl/doc/extensions/Reduction/Reduction.md:44]: But it is a minor thing not worth arguing about. But I do think it would be better if we at least say that `has_known_identity<O,T>` is defined by whether `known_identity<O,T>::identity` exists. That means one can implement it using `is_detected` and in the future (when we add support for user to provide custom `known_identity`) users don't need to provide also `has_known_identity`.

Comment (Pennycook) [reply sycl/doc/extensions/Reduction/Reduction.md]: Good idea.  I've made the change in 0136df20a937.  I swapped the order in which the trait classes are introduced, so that `has_known_identity` can be defined in terms of `known_identity`, and tried to clarify that `known_identity::value` only exists if the value is known.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (rarutyun) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:195]: It seems like `AccumulatorT` is the the fundamental C++ type only (because it's used in `std::numeric_limits`).

Two questions basing on that:
* What about custom types?
* If only fundamental types should be supported, can we introduce the `integral_constant` semantics for `known_identity` API as I suggested for `has_known_identity`?

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (rarutyun) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp:182]: I checked our ``enable_if_t`` implementation. It is the same as in C++ standard. So, there should not be any problem. If you need help we may have a call or IM

File: sycl/doc/extensions/Reduction/Reduction.md
Comment (Pennycook) [reply sycl/doc/extensions/Reduction/Reduction.md:44]: All of the traits in C++ are defined as a `struct`, aren't they?  Is there a good reason that `known_identity` should be different?

I'm not sure what you're imagining when aligning with `integral_constant`.  In the general case, I don't think we can guarantee that the identity value is something that can be passed as a template parameter (e.g. it may be a floating-point type).

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Makes sense.  Is what I've implemented in 6ca8965b32cba6475ca2e457b4026d91b14fa20d what you expected?

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp:194]: This is a short-term fix.  Our reduction implementation currently only detects identity values automatically for fundamental C++ types (and `half`), so that's all we've decided to cover with the traits for now.

Eventually we want developers to be able to declare the identity value for their own types and function object types.  The current thinking is that for combinations of fundamental types and known function object types (e.g. `std::plus<>`) the traits will continue to work as implemented here, but for anything else they'll check the function object for something like an `identity` member.

There are still some open questions about exactly what this should look like for transparent functors and what additional checks are needed (e.g. if we can determine an identity for `T1` and `T1` is convertible to `T2`, is it safe to assume that the identity for `T1` can be used?).  These are important questions, but I don't think the first implementation of the traits should be blocked by them.

File: sycl/include/CL/sycl/half_type.hpp
Comment (rolandschulz) [inline sycl/include/CL/sycl/half_type.hpp:91]: I didn't realize all is fwd-declared and therefore
```suggestion
  friend std::numeric_limits<half_impl::half>;
```
or
```suggestion
  friend std::numeric_limits<sycl::half>;
```
works too. Do we prefer one of these shorter options?

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp
Comment (rarutyun) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Yes, it is.

**Comment by Pennycook** [comment]
@rolandschulz: Are you still working on fixing the broken cases here?

File: sycl/include/CL/sycl/half_type.hpp
Comment (Pennycook) [reply sycl/include/CL/sycl/half_type.hpp:91]: I have no preference.

**Comment by bader** [comment]
We also need resolve conflicts with https://github.com/intel/llvm/pull/2624.

File: sycl/include/CL/sycl/half_type.hpp
Comment (rolandschulz) [inline sycl/include/CL/sycl/half_type.hpp:29]: Why is there the sycl-include-gnu11.cpp test which checks it compiles with C++11? I thought we only support compiling as >=C++14.

Comment (Pennycook) [reply sycl/include/CL/sycl/half_type.hpp:29]: I think the RT library is C++14, but the headers are C++11.

**Comment by Pennycook** [comment]
Thanks for the fixes, @rolandschulz.  Opening up to review now.

File: sycl/include/CL/sycl/half_type.hpp
Comment (v-klochkov) [reply sycl/include/CL/sycl/half_type.hpp:29]: @bader , @romanovvlad - Can't we assume C++14 or newer in SYCL header files?

Comment (bader) [reply sycl/include/CL/sycl/half_type.hpp:29]: I don't have a simple "yes/no" answer, but I'm aware of a couple of factors restricting usage of C++ features:
- SYCL specification: SYCL-1.2.1 relies on C++11 and SYCL-2020 relies on C++17 features. It seems to me that DPC++ is not going to support "strict SYCL-1.2.1" mode where C++14 (or later) features are not supported.
- Runtime library ABI compatibility (see the note in our [contribution guide](https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#development)). It's hard to say if using C++14 features in SYCL headers breaks runtime ABI, but it possible and should be prevented by regression tests.

Comment (pvchupin) [reply sycl/include/CL/sycl/half_type.hpp:29]: We have C++ version documented at:
https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#c-standard
but it doesn't answer whether C++11 mode will be functional or broken.

I think it would be good to know and catch the case when we really want to break C++11 compatibility and have some discussion on that. Maybe that's what sycl-include-gnu11.cpp test is doing?

Comment (rolandschulz) [reply sycl/include/CL/sycl/half_type.hpp:29]: The test only checks that the header can be included. But not that any template in the header can be instantiated in C++11 mode. The current version of this PR passes all tests . But trying to instantiate known_identity for half in C++11 will give a compiler error (not tested).

It shouldn't have any ABI issues to use any C++14 features in the header. The problematic release for ABI issues is C++11 (e.g. string). There weren't any such things in C++14.

Comment (v-klochkov) [reply sycl/include/CL/sycl/half_type.hpp:29]: C++14 is required now. The doc changes have been merged 10 days ago: https://github.com/intel/llvm/pull/3053

**Review by v-klochkov — APPROVED** [review]
Looks good to me. I suggest merging this patch as is now.

A bit later I'll fix the c++11 LIT test (the fix in that LIT test is included into https://github.com/intel/llvm/pull/3123) and also will remove the macro _CPP14_CONSTEXPR .

@Pennycook please approve the fix if you agree with the changes and the idea to merge the fix in the current state.

File: sycl/include/CL/sycl/half_type.hpp
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/half_type.hpp:163]: Code style: A space was removed between `Data(rhs)` and `{}` - I wonder why it was done, because it kind of contradicts the styl of lines 94 and 156

**Review by AlexeySachkov — APPROVED** [review]
`half_type.hpp` changes look good to me, one minor code style comment

File: sycl/include/CL/sycl/half_type.hpp
Comment (v-klochkov) [reply sycl/include/CL/sycl/half_type.hpp:163]: I think those spaces/etc are strictly controlled by clang-format.

**Comment by Pennycook** [comment]
> Looks good to me. I suggest merging this patch as is now.
> 
> A bit later I'll fix the c++11 LIT test (the fix in that LIT test is included into #3123) and also will remove the macro _CPP14_CONSTEXPR .
> 
> @Pennycook please approve the fix if you agree with the changes and the idea to merge the fix in the current state.

Apparently GitHub doesn't let people approve their own PRs, but I agree with your proposal.  Thanks for picking up the required LIT test and macro changes.

File: sycl/include/CL/sycl/half_type.hpp
Comment (AlexeySachkov) [reply sycl/include/CL/sycl/half_type.hpp:163]: Fair enough, but I still wonder why is code style different?

Comment (Pennycook) [reply sycl/include/CL/sycl/half_type.hpp:163]: I might be wrong about this, but I think we only check clang-format has been applied to the lines that change.  There may be lines (or complete files) that haven't been formatted by clang-format, and they won't be updated unless they change.

Comment (AlexeySachkov) [reply sycl/include/CL/sycl/half_type.hpp:163]: > but I think we only check clang-format has been applied to the lines that change.

That is correct, but from what I see on "Files changed" page is that both lines 94 and 163 were changed and to me they seem to have inconsistent formatting - that's what I was wondering about

**Pull request merged**
